ssl/quic/qlog.c

```diff
@@ -0,0 +1,696 @@
+
+    if (p == end) {
+        lex->p          = end;
+        lex->term_end   = end;
+        return 0;
+    }
+
+    for (term_end = p; !is_term_sep_ws(*term_end) && term_end < end; ++term_end);
+
+    lex->p          = p;
+    lex->term_end   = term_end;
+    return 1;
+}
+
+static int lex_eot(struct lexer *lex)
+{
+    return lex->p == lex->term_end;
+}
+
+static int lex_peek_char(struct lexer *lex)
+{
+    return lex_eot(lex) ? -1 : *lex->p;
+}
+
+static int lex_skip_char(struct lexer *lex)
+{
+    if (lex_eot(lex))
+        return 0;
+
+    ++lex->p;
+    return 1;
+}
+
+static int lex_match(struct lexer *lex, const char *s, size_t s_len)
+{
+    if ((size_t)(lex->term_end - lex->p) != s_len)
+        return 0;
+
+    if (memcmp(lex->p, s, s_len))
+        return 0;
+
+    return 1;
+}
+
+static void lex_get_rest(struct lexer *lex, const char **str, size_t *str_l)
+{
+    *str    = lex->p;
+    *str_l  = lex->term_end - lex->p;
+}
+
+static int lex_extract_to(struct lexer *lex, char c,
+                          const char **str, size_t *str_l)
+{
+    const char *p = lex->p, *term_end = lex->term_end, *s;
+
+    for (s = p; s < term_end && *s != c; ++s);
+    if (s == term_end)
+        return 0;
+
+    *str    = p;
+    *str_l  = term_end - s;
+    lex->p  = ++s;
+    return 1;
+}
+
+static int ossl_unused filter_match_event(const char *cat, size_t cat_l,
+                                          const char *event, size_t event_l,
+                                          const char *expect_cat,
+                                          const char *expect_event)
+{
+    size_t expect_cat_l = strlen(expect_cat);
+    size_t expect_event_l = strlen(expect_event);
+
+    if ((cat != NULL && cat_l != expect_cat_l)
+        || (event != NULL && event_l != expect_event_l)
+        || (cat != NULL && memcmp(cat, expect_cat, expect_cat_l))
+        || (event != NULL && memcmp(event, expect_event, expect_event_l)))
+        return 0;
+
+    return 1;
+}
+
+/*
+ * enabled: event enablement bitmask Array of size NUM_ENABLED_W.
+ * add: 1 to enable an event, 0 to disable.
+ * cat: Category name/length. Not necessarily zero terminated.
+ *      NULL to match any.
+ * event: Event name/length. Not necessarily zero terminated.
+ *        NULL to match any.
+ */
+static void filter_apply(size_t *enabled, int add,
+                         const char *cat, size_t cat_l,
+                         const char *event, size_t event_l)
+{
+    /* Find events which match the given filters. */
+# define QLOG_EVENT(e_cat, e_name)                          \
+    if (filter_match_event(cat, cat_l, event, event_l, \
+                                #e_cat, #e_name))           \
+        bit_set(enabled, QLOG_EVENT_TYPE_##e_cat##_##e_name, add);
+# include "internal/qlog_events.h"
+# undef QLOG_EVENT
+}
+
+static int lex_fail(struct lexer *lex, const char *msg)
+{
+    /*
+     * TODO(QLOG): Determine how to print log messages about bad filter strings
```

Comment:
What is the plan for resolving this? Do we have an issue?

Reply:
What would you prefer? Unclear what our strategy should be here. Not sure we ever really just emit output because an application which links libssl is run with an environment variable set.